draft-ietf-tls-tlsflags.xml

```diff
@@ -131,12 +131,13 @@
             <t> A server that supports this extension and also supports at least one of the flag-type features
                 that use this extension and that were declared by the ClientHello extension SHALL send this
                 extension with the intersection of the flags it supports with the flags declared by the client.
-                The intersection operation MAY be implemented as a bitwise AND. The server may need to send two
-                tls_flags extensions, one in the ServerHello and the other in the EncryptedExtensions message.
+                The intersection operation MAY be implemented as a bitwise AND. The server may need to send more than one
+                tls_flags extensions, up to one in each of the messages that may carry extensions: ServerHello, 
+                EncryptedExtensions, Certificate, and NewSessionTicket.
                 It is up to the document for the specific feature to determine whether support should be
-                acknowledged in the ServerHello or the EncryptedExtensions message.</t>
+                acknowledged in the ServerHello, in the EncryptedExtensions, or in one of the other messages.</t>
             <t> A server MUST NOT indicate support for any flag-type feature not previously indicated by the 
-                client. It MUST NOT include this extension in either message (ServerHello or EncryptedExtensions)
+                client. It MUST NOT include this extension in any of the four messages 
```

Comment:
This seems a bit problematic because it means that the server can't for instance, indicate flags in NST if the client hasn't indicated the extension in CH. Unfortunately, the draft currently forbids sending an empty extension in CH. I suggest we relax this text to make clear it's only about "response" extensions

Reply:
The relevant text in from [RFC8446](https://tools.ietf.org/html/rfc8446#section-4.2) is: "The server MAY also send unsolicited extensions in the NewSessionTicket, though the client does not respond directly to these."

Reply:
I might relax this as follows:

"With the exception of NST, a server MUST NOT indicate support for any flag-type feature not previously indicated by the client in CH, and MUST NOT include the tls_flags extension in any message if there is no appropriate flag-type to indicate. Servers MAY send unsolicited tls_flag flag-type features in NST, though the client does not need to consume or otherwise respond to these."

Reply:
I'm not sure that I agree.
In any of the server-side messages that can include an extension, the extension is telling the client either about something that the server is going to do, or something that the server wants the client to do.  There are not going to be any opportunities for the client to indicate support (or lack thereof), so it's better for those to be pre-announced.

So if we have some extension in the NST that, perhaps, tells the Client how to handle the session ticket, it makes sense to know in advance that the Client knows what that flag means.  Although this is not available for regular extensions, there's no reason to deny us this in flags.  What's the harm of requiring the Client to pre-announce that it supports this NST extension flag?

Reply:
I would say rather what's the value in it?

My mental model is that flags are simply compressed versions of extensions and so should follow the same rules. that means that you should be able to offer unsolicited flags in any place you could offer an unsolicited extension, which includes NST and CR.

Reply:
I'm fine with that outcome, but if we require the Client to indicate support then the Server can know that the Client will understand the extension. If we don't it doesn't. And with the flags extension, indicating support is cheap.
We might want to take it to the list.

Reply:
@yoavnir please do so. We need to get this PR moving!